doc/link_event_damping/Link-event-damping-HLD.md

```diff
@@ -0,0 +1,313 @@
+
+###  1.6. <a name='ArchitectureDesign'></a>Architecture Design
+
+Today, in SONiC when link up/down event notifications are triggered by vendor SAI layer, a callback from SAI goes into SyncD and SyncD generates notification in ASIC DB.
+All the consumers (practically speaking PortsOrch) listen to this notification and update the state for upstream consumers.
+
+![current link events path](current_link_events_path.jpg)
+
+In this proposal, implementation of link event damping will be in the SyncD where it will hook into the port state change notification callback from vendor SAI to SyncD and intercept the notifications.
+In SyncD, this will provide an entry point for proposed implementation. Link event damping implementation will be in the main selectable loop of the SyncD thread.
+
+An alternate approach can be to add the implementation of the feature in the vendor SAI layer.
+
+We will discuss both the approaches in next section.
+
+
+###  1.7. <a name='High-LevelDesign'></a>High-Level Design
+
+#### 1.7.1. <a name='SyncDImplementation'></a>SyncD Implementation
+
+As mentioned in the previous section, the proposal is to implement link event damping in the SyncD where it can hook into the port state change notification callback from the vendor SAI.
+
+
+![syncd link event damping path](syncd_link_event_damping_path.jpg)
+
+In this design, link-event damping related configurations will be pushed to CONFIG DB per logical interface.
+
+To push these per interface configuration down to the stack, Orch Agent will be modified such that Orch Agent translates per-interface config to notifications to SAI - similar to apply view - using the switch object API - `set_switch_attribute`. This approach requires adding a redis switch attribute to `sai_redis_switch_attr_t` for hold-time config. Since hold-time configuration is consumed in SONiC stack itself, it is being treated as switch (aka redis switch) attribute thus using the switch object API to pass down the configuration from Orch Agent to SyncD.
+
+SyncD will monitor these configs, consume them and apply them on the appropriate interface. Syncd will maintain a link event damper for each interface and apply the required damping before sending the interface state change notification to any consumers.
+
+When an interface is removed, SyncD will remove the corresponding link event damper by monitoring the `SAI_COMMON_API_REMOVE` for `SAI_OBJECT_TYPE_PORT` object type.
+
+#####  1.7.1.1. <a name='NetdevPort'></a>Link Event damping on Netdev port
+
+When a physical port’s link state changes, its corresponding netdev interface is expected to change its oper state too. SONiC has an existing SAI hostif API that OA uses to set the host interface’s oper status using SAI attribute `SAI_HOSTIF_ATTR_OPER_STATUS`. By effect of this proposal OA receives only the damped link state, netdev port’s state too will be damped accordingly.
+
+
+#####  1.7.1.2. <a name='Example'></a>Example
+
```

Comment:
Aren't examples below applicable for both proposed implementations ?

Reply:
not valid anymore.